packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts

```diff
@@ -2240,185 +2463,205 @@ class ObjectsInRedisClient {
             callback = options;
             options = null;
         }
-        utils.checkObjectRights(this, null, null, options, 'list', (err, options) => {
+        utils.checkObjectRights(this, null, null, options, 'list', err => {
             if (err) {
                 return tools.maybeCallbackWithRedisError(callback, err);
             } else {
-                return this._subscribe(pattern, options, this.sub, callback);
+                return this._subscribe(pattern, true, callback);
             }
         });
     }
 
-    subscribeUserAsync(pattern, options) {
-        return /** @type {Promise<void>} */ (
-            new Promise((resolve, reject) =>
-                this.subscribeUser(pattern, options, err => (err ? reject(err) : resolve()))
-            )
+    subscribeUserAsync(pattern: string, options: CallOptions): Promise<void> {
+        return new Promise<void>((resolve, reject) =>
+            this.subscribeUser(pattern, options, err => (err ? reject(err) : resolve()))
         );
     }
 
-    _unsubscribe(pattern, options, subClient, callback) {
+    private async _unsubscribe(pattern: string | string[], asUser: boolean): Promise<void> {
+        const subClient = asUser ? this.sub : this.subSystem;
         if (!subClient) {
-            return tools.maybeCallbackWithRedisError(callback, ERRORS.ERROR_DB_CLOSED);
+            throw new Error(ERRORS.ERROR_DB_CLOSED);
         }
         if (Array.isArray(pattern)) {
-            let count = pattern.length;
-            pattern.forEach(pattern => {
-                this.log.silly(this.namespace + ' redis punsubscribe ' + this.objNamespace + pattern);
-                subClient.punsubscribe(this.objNamespace + pattern, err => {
-                    if (!err && subClient.ioBrokerSubscriptions[this.objNamespace + pattern] !== undefined) {
-                        delete subClient.ioBrokerSubscriptions[this.objNamespace + pattern];
-                    }
-                    !--count && typeof callback === 'function' && callback(err);
-                });
-            });
-        } else {
-            this.log.silly(this.namespace + ' redis punsubscribe ' + this.objNamespace + pattern);
-            subClient.punsubscribe(this.objNamespace + pattern, err => {
-                if (!err && subClient.ioBrokerSubscriptions[this.objNamespace + pattern] !== undefined) {
-                    delete subClient.ioBrokerSubscriptions[this.objNamespace + pattern];
+            for (const _pattern of pattern) {
+                this.log.silly(`${this.namespace} redis punsubscribe ${this.objNamespace}${_pattern}`);
+                await subClient.punsubscribe(this.objNamespace + _pattern);
+                const subscriptions = asUser ? this.userSubscriptions : this.systemSubscriptions;
+                if (subscriptions[this.objNamespace + _pattern] !== undefined) {
+                    delete subscriptions[this.objNamespace + _pattern];
                 }
```

Comment:
Sounds like this could be turned into a Map

Reply:
yes it can, I also thought about it recently, I would do it in another PR because the same structure exists in states client.